src/ipfs.rs

```diff
@@ -17,36 +38,104 @@ pub type KeplerParams = DefaultParams;
 //     type Hashes = Code;
 // }
 
-pub type Ipfs = OIpfs<KeplerParams>;
+pub type Ipfs = OIpfs<Types>;
 pub type Block = OBlock<KeplerParams>;
+pub type Swarm = TSwarm<Types>;
+
+pub async fn create_ipfs<'l, I>(
+    id: String,
+    dir: &'l Path,
+    keypair: Keypair,
+    allowed_peers: I,
+) -> Result<(Ipfs, impl Future<Output = ()>, Receiver<BehaviourEvent>)>
+where
+    I: IntoIterator<Item = PeerId> + 'static,
+{
+    let ipfs_path = dir.join("ipfs");
+    std::fs::create_dir(&ipfs_path)?;
+    let ipfs_opts = IpfsOptions {
+        ipfs_path,
+        keypair,
+        bootstrap: vec![],
+        mdns: false,
+        kad_protocol: Some("/ipfs/lan/kad/1.0.0".to_string()),
+        listening_addrs: vec![multiaddr!(P2pCircuit)],
+        span: None,
+    };
+
+    let (sender, receiver) = std::sync::mpsc::sync_channel::<BehaviourEvent>(100);
+    let behaviour = Behaviour::new(sender);
+
+    let (transport_builder, relay_behaviour) = TransportBuilder::new(ipfs_opts.keypair.clone())?
+        .or(MemoryTransport::default())
+        .relay();
+
+    let transport = transport_builder
+        .map_auth()
+        .map(crate::transport::auth_mapper(allowed_peers))
+        .build();
+
+    let (ipfs, ipfs_task) =
+        UninitializedIpfs::<Types>::new(ipfs_opts, transport, Some(relay_behaviour))
+            .with_extended_behaviour(behaviour)
+            .start()
+            .await?;
+
+    Ok((ipfs, ipfs_task, receiver))
+}
+
+struct Content {
+    parts: Vec<Cid>,
+}
 
 #[rocket::async_trait]
 impl ContentAddressedStorage for Ipfs {
     type Error = anyhow::Error;
     async fn put(&self, content: &[u8], _codec: SupportedCodecs) -> Result<Cid, Self::Error> {
-        // TODO find a way to stream this better? (use .take with max block size?)
-        // TODO impl support for chunking with linked data (e.g. use IpldCodec)
-        let block = Block::encode(RawCodec, Code::Blake3_256, content)?;
-        self.insert(&block)?;
-        self.alias(block.cid().to_bytes(), Some(block.cid()))?;
-        Ok(*block.cid())
+        let parts: Vec<Ipld> = content
+            .chunks(KeplerParams::MAX_BLOCK_SIZE)
+            .map(Vec::from)
+            .map(Ipld::Bytes)
+            .collect();
+
+        let dag = Ipld::List(parts);
+        let cid = self.put_dag(dag).await?;
+        self.insert_pin(&cid, true).await?;
+        Ok(cid)
```

Comment:
actually, this will fail for `content.len() > KeplerParams::MAX_BLOCK_SIZE`. `self.put_dag` will turn all of `dag` (a list of IPLD byte objects of max block size, not a list of cids) into a single block of size N * max size